src/HttpRequest.cc

```diff
@@ -543,7 +537,7 @@ HttpRequest *
 HttpRequest::FromUrl(const char * url, const MasterXaction::Pointer &mx, const HttpRequestMethod& method)
 {
     std::unique_ptr<HttpRequest> req(new HttpRequest(mx));
-    if (req->url.parse(method, url)) {
+    if (req->url.parse(method, SBuf(url))) {
```

Comment:
Please move this performance regression into the callers. Some of them actually convert SBuf to c-string just to call this function! If you think that there are too many callers that cannot use SBuf without significant changes, then you can add a temporary FromUrlXXX() method that accepts a c-string.

Reply:
There are 13 c-string callers and 2 SBuf - both of which are in uncommon code paths. I tried that in an earlier experiment, it caused libsbuf.la dependencies in a number of the unit tests and other library objects so there were quite a few more Makefile changes and re-shuffling of linker lists needed to make it work. So I left it as-is to avoid bloating this PR patch any more.

Reply:
Done.

Reply:
A lot of recently added FromUrlXXX() callers can (and, hence, should) use SBuf without significant changes. The whole point of adding a FooXXX() function is to mark places that cannot be easily fixed, that require some substantial analysis, widespread changes, other serious development work, etc. Sprinkling Squid code with FooXXX() calls where Foo() can be used without serious additional work makes the code worse, not better.

Specifically, most test cases where FromUrlXXX() was added should probably use SBuf instead of a c-string. The call with a `local_urlres` parameter can be easily converted as well AFAICT. I did not check all the callers because the ball is still on your side, but please keep this comment in mind as you prepare this PR for the next round of reviews so that we can reduce the number of review iterations required to polish this code.

Reply:
Now that this is a security fix I/we need to place line count minimizing over polish changes. The use of FromUrlXXX has already doubled the touched line count and is pushing it towards the realms of not being able to backport into the 3.x which need it as a security fix. I only went with it because of your worry about a serious performance regression from c_str().

The local_res is going to become an AnyP::Uri shortly in another PR. I already know that touching it in any way is far wider reaching than this PR should handle. Similar with the FTP caller. Probably the others as well. If there is still a performance issue you can see still in the main HTTP pathway caller(s) I will consider them, but not the rarely used callers.

Reply:
I am resolving this change request because it combined two specific areas: test cases and local_urlres, and I agree that we can leave local_urlres as is for now. I will create another, more context-specific change request for test cases.